src/main/scala/cognite/spark/v1/SdkV1Relation.scala

```diff
@@ -123,14 +116,37 @@ abstract class SdkV1Relation[A <: Product, I](config: RelationConfig, shortName:
       T <: UpdateByExternalId[R, U, IO] with Create[R, C, IO]](
       existingExternalIds: Seq[String],
       resourceCreates: Seq[C],
-      resource: T)(implicit transform: Transformer[C, U]): IO[Unit] = {
+      resource: T,
+      doUpsert: Boolean)(implicit transform: Transformer[C, U]): IO[Unit] = {
     val (resourcesToUpdate, resourcesToCreate) = resourceCreates.partition(
       p => if (p.externalId.isEmpty) { false } else { existingExternalIds.contains(p.externalId.get) }
     )
-    val create = if (resourcesToCreate.isEmpty) { IO.unit } else {
-      resource.create(resourcesToCreate)
+    val create = if (resourcesToCreate.isEmpty) {
+      IO.unit
+    } else {
+      resource
+        .create(resourcesToCreate)
+        .flatTap { _ =>
+          IO {
+            if (config.collectMetrics) {
+              itemsCreated.inc(resourcesToCreate.size)
+            }
+          }
+        }
+        .handleErrorWith {
+          case e: CdpApiException if doUpsert && e.code == 409 =>
+            val existingExternalIds =
+              e.duplicated.get.map(j => j("externalId").get.asString.get)
```

Comment:
Could this be improved by parsing the Json directly to a case class, either here or in the SDK?